src/map/script.c

```diff
@@ -5685,24 +5693,64 @@ BUILDIN(goto)
 	return true;
 }
 
-/*==========================================
- * user-defined function call
- *------------------------------------------*/
+/// User-defined function call and subroutine call (Updated by [Cretino]) (Thanks to [FlavioJS])
+/// 'callfunc' support labels in the current npc "::LabelName"
+/// 'callfunc' support labels in other npcs "NpcName::LabelName"
 BUILDIN(callfunc)
 {
-	int i, j;
-	struct script_retinfo* ri;
-	struct script_code* scr;
-	const char* str = script_getstr(st,2);
+	int i, j, k, pos = -1;
+	struct script_retinfo *ri;
+	struct script_code *scr = NULL;
 	struct reg_db *ref = NULL;
+	struct npc_data *nd = NULL;
+	struct script_data *data = script_getdata(st, 2);
 
-	scr = (struct script_code*)strdb_get(script->userfunc_db, str);
-	if( !scr )
-	{
-		ShowError("script:callfunc: function not found! [%s]\n", str);
-		st->state = END;
-		return false;
-	}
+	if( !data_islabel(data) && !data_isfunclabel(data) ) {
+		const char *str = script_getstr(st, 2);
+		scr = (struct script_code*)strdb_get(script->userfunc_db, str);
+
+		if (!scr) {
+			if (stristr(str, "::")) {
+				char labelname[NAME_LENGTH], npcname[NAME_LENGTH];
+
+				memset(labelname, '\0', sizeof(labelname));
+				memset(npcname, '\0', sizeof(npcname));
+
+				if ( sscanf(str, "::%s", labelname) == 1 || sscanf(str, "%[^:]::%s", npcname, labelname) == 2 ) {
+					if ( !labelname[0] && !npcname[0] ) {
+						ShowError("script:callfunc: label not found! [%s]\n", str);
+						st->state = END;
+						return false;
+					}
+
+					nd = (npcname[0]) ? npc->name2id(npcname) : map->id2nd(st->oid);
+
+					if (!nd) {
+						ShowError("script:callfunc: no npc attached for label! [%s]\n", str);
+						st->state = END;
+						return false;
+					}
+
+					ARR_FIND(0, nd->u.scr.label_list_num, k, (strcmp(nd->u.scr.label_list[k].name, labelname) == 0));
+
+					if (k == nd->u.scr.label_list_num) {
+						ShowError("script:callfunc: label not found! [%s]\n", str);
+						st->state = END;
+						return false;
+					}
+				} else {
+					ShowError("script:callfunc: label not found! [%s]\n", str);
+					st->state = END;
+					return false;
+				}
+			} else {
+				ShowError("script:callfunc: argument is not a label or function not found! [%s]\n", str);
+				st->state = END;
+				return false;
+			}
+		}
+	} else
+		pos = script_getnum(st, 2);
 
 	ref = (struct reg_db *)aCalloc(sizeof(struct reg_db), 2);
```

Comment:
The previous implementation only allocates two instances of reg_db in the case of function calls (there's no need to save a reference to the `.vars` if it's a subroutine call inside the same NPC) - compare this line in buildin_callsub and buildin_callfunc.

This means that after this commit, callsub becomes slower and more memory hungry (because it causes an extra DBMap to be created). Perhaps it's worth separating the two cases.